core/tests/LowLevelTests/Op.hs

```diff
@@ -1,60 +1,81 @@
+  testGroup
+    "Synchronous unit tests of low-level Op interface"
+    [testCancelFromServer]
 
 testCancelFromServer :: TestTree
 testCancelFromServer =
   testCase "Client/Server - client receives server cancellation" $
-  runSerialTest $ \grpc ->
-    withClientServerUnaryCall grpc $
-    \(Client{..}, Server{..}, ClientCall{..}, sc@ServerCall{..}) -> do
-      serverCallCancel sc StatusPermissionDenied "TestStatus"
-      clientRes <- runOps unsafeCC clientCQ clientRecvOps
-      case clientRes of
-        Left x -> error $ "Client recv error: " ++ show x
-        Right [_,_,OpRecvStatusOnClientResult _ code _details] -> do
-          code @?= StatusPermissionDenied
-          return $ Right ()
-        wrong -> error $ "Unexpected op results: " ++ show wrong
-
+    runSerialTest $ \grpc ->
+      withClientServerUnaryCall grpc $
+        \(Client {..}, Server {..}, ClientCall {..}, sc@ServerCall {..}) -> do
+          serverCallCancel sc StatusPermissionDenied "TestStatus"
+          clientRes <- runOps unsafeCC clientCQ clientRecvOps
+          case clientRes of
+            Left x -> error $ "Client recv error: " ++ show x
+            Right [_, _, OpRecvStatusOnClientResult _ code _details] -> do
+              code @?= StatusPermissionDenied
+              return $ Right ()
+            wrong -> error $ "Unexpected op results: " ++ show wrong
 
 runSerialTest :: (GRPC -> IO (Either GRPCIOError ())) -> IO ()
 runSerialTest f =
-  withGRPC f >>= \case Left x -> error $ show x
-                       Right () -> return ()
+  withGRPC f >>= \case
+    Left x -> error $ show x
+    Right () -> return ()
 
-withClientServerUnaryCall :: GRPC
-                             -> ((Client, Server, ClientCall,
-                                  ServerCall ByteString)
-                                 -> IO (Either GRPCIOError a))
-                             -> IO (Either GRPCIOError a)
+withClientServerUnaryCall ::
+  GRPC ->
+  ( ( Client,
+      Server,
+      ClientCall,
+      ServerCall ByteString
+    ) ->
+    IO (Either GRPCIOError a)
+  ) ->
+  IO (Either GRPCIOError a)
 withClientServerUnaryCall grpc f = do
   withClient grpc clientConf $ \c -> do
     crm <- clientRegisterMethodNormal c "/foo"
-    withServer grpc serverConf $ \s ->
-      withClientCall c crm 10 $ \cc -> do
+    withServer grpc serverConf $ \s -> do
+      ccVar <- newEmptyMVar
+      bracket newEmptyMVar (\v -> putMVar v ()) $ \finished -> do
+        _ <- forkIO $
+          void $
+            withClientCall c crm 10 $ \cc -> do
+              putMVar ccVar cc
+              -- NOTE: We need to send client ops here or else `withServerCall` hangs,
+              -- because registered methods try to do recv ops immediately when
+              -- created. If later we want to send payloads or metadata, we'll need
+              -- to tweak this.
+              _clientRes <- runOps (unsafeCC cc) (clientCQ c) clientEmptySendOps
+              takeMVar finished
```

Comment:
It seems like this `takeMVar` will always throw a `BlockedIndefinitelyOnMVar` exception.  Is that the intention?

Reply:
I don’t think it does. the `bracket` call puts the mvar We could probably make the whole logic here much nice with `async`. Just didn’t seem worth it just for a test.

Reply:
What I mean is that the `bracket` call only invokes `putMVar` as part of its `finally` clause.  In other words, my understanding is that the `putMVar` is only called when the bracketed code finishes or throws an exception, but that would only occur after the `takeMVar` call

Reply:
the bracketed code block is on a different thread so it can occur before we get here. That said looking at this, I can’t remember why I didn’t want to keep this thread alive in the first place. I think we might be able to just delete this. I never found the time to clean this up which is why I didn’t upstream it myself.